### PR TITLE
Obstacle detection

### PR DESCRIPTION
**Pull request by vvasco**
This PR introduces the `obstacleDetector` module, which:
- clusters data received from front and back lasers, by means of Euclidean distance using the `OpenCv` function [`partition`](https://docs.opencv.org/3.4/d5/d38/group__core__cluster.html#ga2037c989e69b499c1aa271419f3a9b34);
- reject clusters with few points (these are considered outliers);
- sends a `stop` to `navController` when the robot reaches a threshold distance (`1.5m`) from the closest obstacle.

The following pics show an example of how the module works in simulation.
The `gazebo` scenario includes the robot with two humans, one in front and one at the back of the robot. 
The `yarplaserscannergui` on the left shows measurements from the lasers (dots).
The `yarpview` on the right shows the same measurements, highlighting points belonging to different clusters with different colors. Also clusters rejected are not shown.  

### Clusters creation and rejection
This picture shows that:
- legs form a single cluster: the purple from the front laser and the pink from the back laser;
- only points at `~4m` from the front and `~2m` from the back form valid clusters, while the remaining points are rejected.
  
clusters-rejection

### Stopping navigation if obstacle detected
If we move the robot along the `x` direction up to `4.0m`, `obstacleDetector` detects an obstacle along the way and stops `navController` at `~1.5m` from the obstacle:

clusters-front-opt

### CODE REVIEW

File: modules/obstacleDetector/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+    double getMinDistance(const double &d1, const double &d2, string &s)
+    {
+        if (d1<d2)
+        {
+            s="front-laser";
+            return d1;
+        }
+        else
+        {
+            s="back-laser";
```
Comment (pattacini) [inline modules/obstacleDetector/src/main.cpp:448]: I would use the term `rear-laser` in lieu of `back-laser`.

File: modules/obstacleDetector/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+using namespace yarp::sig;
+using namespace yarp::math;
+using namespace cv;
+
+/****************************************************************/
+class ObstDetector : public RFModule
+{
+    double period{0.0};
+    string module_name{"obstacleDetector"};
+    string robot{"SIM_CER_ROBOT"};
```
Comment (pattacini) [inline modules/obstacleDetector/src/main.cpp:53]: I would rather initialize this variable with the tag of a physical robot.

File: modules/obstacleDetector/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+            return sqrt((a.x-b.x)*(a.x-b.x) + (a.y-b.y)*(a.y-b.y)) < threshold;
+        }
+    };
+
+    /****************************************************************/
+    bool configure(ResourceFinder& rf) override
+    {
+        period=rf.check("period",Value(0.1)).asDouble();
+        module_name=rf.check("name",Value("obstacleDetector")).asString();
+        robot=rf.check("robot",Value("SIM_CER_ROBOT")).asString();
```
Comment (pattacini) [inline modules/obstacleDetector/src/main.cpp:108]: Same comment as above.

File: modules/obstacleDetector/src/main.cpp (hunk 1 above)
Comment (pattacini) [reply modules/obstacleDetector/src/main.cpp:448]: Along the same line, I would replace `back` with `rear` everywhere.

Comment (vvasco) [reply modules/obstacleDetector/src/main.cpp:448]: I used `back` to be consistent with the name of the port opened by the laser, which is `/SIM_CER_ROBOT/laser/back:o`.

Comment (pattacini) [reply modules/obstacleDetector/src/main.cpp:448]: Ok, so the "typo" is upstream 😉 
It's up to you then.

File: modules/obstacleDetector/src/main.cpp (hunk 3 above)
Comment (vvasco) [reply modules/obstacleDetector/src/main.cpp]: Done.

File: modules/obstacleDetector/src/main.cpp (hunk 2 above)
Comment (vvasco) [reply modules/obstacleDetector/src/main.cpp]: Done.

File: modules/obstacleDetector/src/main.cpp (hunk 1 above)
Comment (vvasco) [reply modules/obstacleDetector/src/main.cpp]: I went for `rear`.

**Pull request merged**